File: LeadersArray.c
#include <stdio.h>
#include <stdlib.h>

int* leadersArray(int n, int arr[], int* resSize) {
    // Cấp phát bộ nhớ cho mảng kết quả, tối đa n phần tử
    int* res = (int*)malloc(n * sizeof(int));
    int count = 0;

    // Khởi tạo phần tử cuối cùng là phần tử dẫn đầu đầu tiên
    int maxRight = arr[n - 1];
    res[count++] = maxRight;  // Thêm phần tử cuối cùng vào mảng kết quả

    // Duyệt mảng từ phải qua trái
    for (int i = n - 2; i >= 0; i--) {  // Bắt đầu từ phần tử thứ hai từ cuối
        if (arr[i] >= maxRight) {  // Nếu phần tử hiện tại lớn hơn hoặc bằng maxRight
            maxRight = arr[i];  // Cập nhật maxRight
            res[count++] = arr[i];  // Thêm phần tử vào mảng kết quả
        }
    }

    // Đảo ngược mảng kết quả để duy trì thứ tự ban đầu
    for (int i = 0; i < count / 2; i++) {
        int temp = res[i];
        res[i] = res[count - i - 1];
        res[count - i - 1] = temp;
    }
    *resSize = count;  // Cập nhật kích thước mảng kết quả
    return res;
}

int main() {
    // Nhập số lượng phần tử mảng
    int n;
    printf("Nhập số lượng phần tử trong mảng: ");
    scanf("%d", &n);

    // Nhập mảng
    int arr[n];
    printf("Nhập các phần tử của mảng cách nhau bởi khoảng trắng: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    // Tìm các phần tử dẫn đầu
    int resSize;
    int* result = leadersArray(n, arr, &resSize);

    // In mảng các phần tử dẫn đầu
    printf("Các phần tử dẫn đầu là: ");
    for (int i = 0; i < resSize; i++) {
        printf("%d ", result[i]);
    }
    printf("\n");

    // Giải phóng bộ nhớ đã cấp phát
    free(result);
    return 0;
}

File: LeadersArray.py
def leaders(arr):
    result = []  # Mảng lưu trữ các phần tử dẫn đầu
    n = len(arr)  # Lấy độ dài của mảng
    # Khởi tạo phần tử cuối cùng của mảng là phần tử dẫn đầu đầu tiên
    maxRight = arr[-1]
    # Phần tử cuối cùng luôn là một phần tử dẫn đầu
    result.append(maxRight)
    # Duyệt mảng từ phải qua trái, bắt đầu từ phần tử thứ hai từ cuối
    for i in range(n - 2, -1, -1):  # Duyệt từ n-2 đến 0
        if arr[i] >= maxRight:  # Nếu phần tử hiện tại lớn hơn hoặc bằng maxRight
            maxRight = arr[i]  # Cập nhật maxRight
            result.append(maxRight)  # Thêm phần tử vào mảng kết quả
    # Đảo ngược mảng kết quả để duy trì thứ tự ban đầu
    result.reverse()
    return result  # Trả về mảng các phần tử dẫn đầu

if __name__ == "__main__":
    # Yêu cầu người dùng nhập mảng
    arr = list(map(int, input("Hãy nhập các phần tử của mảng cách nhau bởi khoảng trắng: ").split()))
    # Tìm các phần tử dẫn đầu
    result = leaders(arr)
    # In kết quả
    print("Các phần tử dẫn đầu là:", " ".join(map(str, result)))
